eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/ClusteringWriter.java

```diff
@@ -276,32 +270,59 @@ private void clusterSubmittedVariants(List<? extends SubmittedVariantEntity> sub
                                                               SubmittedVariantEntity.class);
         BulkOperations dbsnpBulkOperations = mongoTemplate.bulkOps(BulkOperations.BulkMode.UNORDERED,
                                                                    DbsnpSubmittedVariantEntity.class);
+        BulkOperations bulkHistoryOperations = mongoTemplate.bulkOps(BulkOperations.BulkMode.UNORDERED,
+                SubmittedVariantOperationEntity.class);
+        BulkOperations dbsnpBulkHistoryOperations = mongoTemplate.bulkOps(BulkOperations.BulkMode.UNORDERED,
+                DbsnpSubmittedVariantOperationEntity.class);
         long numUpdates = 0;
         long numDbsnpUpdates = 0;
         for (SubmittedVariantEntity submittedVariantEntity : submittedVariantEntities) {
             if (submittedVariantEntity.getClusteredVariantAccession() != null) {
                 // no need to update the rs of a submitted variant that already has the correct rs
                 continue;
             }
-            Query query = query(where("_id").is(submittedVariantEntity.getId()));
-            Update update = new Update();
-            update.set(RS_KEY, getClusteredVariantAccession(submittedVariantEntity));
+
+            long rsid = getClusteredVariantAccession(submittedVariantEntity);
+            // Query to update the RSid in submittedVariantEntity
+            Query rs_query = query(where("_id").is(submittedVariantEntity.getId()));
+            Update rs_update = new Update();
+            rs_update.set(RS_KEY, rsid);
+
+            // Query to create the update operation
+            SubmittedVariantOperationEntity update_operation = createUpdateOperations(submittedVariantEntity, rsid);
```

Comment:
either extract the creation of both operations, or none. Currently it's mixing different levels of abstraction.

Reply:
Decided on fewer functions. Happy to discuss if you think it should be the other way around.